### DISCUSSION TITLE
help: how do you make custom queries with javascript sdk?

### DISCUSSION THREAD

**Discussion by jumper385** [pocketbase/pocketbase#579] [Q&A]
I'm having trouble trying to do the following:
1) get order `table` to `location` table data through `asset2order`, `asset2location` tables. How would I achieve this on the javascript SDK?
2) Is there a way to write sql queries using the javascript sdk?

**Comment by ganigeorgiev** [comment]
1. I'm not sure that I understand your example, but in general you can fetch relation records by using the `expand` query parameter:
    ```js
    client.records.getList("example", 1, 50, { expand: "yourRelationFieldName" });
    ```
    Nested relation expansion is also supported using the dot-notation (eg. `yourRelationFieldName.anotherRelation`).
    Note that you have to allow the **view API rule** for each expanded relation(s) collection in order to be able to expand the collection records.
    In v.0.8.0 we will have also support for indirect expansion (aka. expanding relation for collections where you don't have a relation field in the main collection) - https://github.com/pocketbase/pocketbase/issues/312

2. No. You can execute custom DB queries only if you use [PocketBase as framework](https://pocketbase.io/docs/use-as-framework/). The SDK (and the REST API) supports only querying records via the `filter` query parameter:
    ```js
    client.records.getList("example", 1, 50, { filter: "status = 'public' && title != ''" });
    ```

_I'll move the issue to Q&A discussion._

**Reply by corvalan-roberto** [reply]
@ganigeorgiev , 
I downloaded the project from minute 1, it seems like a great thing to me. But I find myself frustrated. I can't relate (Expand) I tried every possible way and from what I see. Many of us come across this topic "I THINK IT DESERVES A CONCRETE EXAMPLE IN THE DOCUMENTATION"

PLEASE!

I do not speak English very well, Sorry!

**Reply by ganigeorgiev** [reply]
@corvalan-roberto What is your data structure? In order to be able to expand a relation:

1. In collection `example`, create a `relation` type field, named "myRelation" and associate it to the collection you want.
2. Go to the **related collection** and make sure that in the related collection API rules the **"view" rule** is configured.
3. From client-side, call `client.records.getList("example", 1, 50, { expand: "myRelation" })`.

I'm not really sure how to explain it better. Could you provide a code sample that shows what you are trying to do?

**Reply by corvalan-roberto** [reply]
> @corvalan-roberto What is your data structure? In order to be able to expand a relation:
> 
> 1. In collection `example`, create a `relation` type field, named "myRelation" and associate it to the collection you want.
> 2. Go to the **related collection** and make sure that in the related collection API rules the **"view" rule** is configured.
> 3. From client-side, call `client.records.getList("example", 1, 50, { expand: "myRelation" })`.
> 
> I'm not really sure how to explain it better. Could you provide a code sample that shows what you are trying to do?

@ganigeorgiev 
Now I was able, I had a conceptual error, in "expand" one has to put the name of the field by which I relate it 🤦‍♂️. Thank you very much for your attention!

Let me congratulate you for this great contribution to the community.!!!

**Comment by ejrua** [comment]
Is there a possible solution in the future?

**Reply by ganigeorgiev** [reply]
@ejrua If you are asking for executing custom SQL via the JS SDK (or REST API) - No. This is not how PocketBase was intended to be used. You can use the `filter` query parameter as shown in the example above.

**Reply by ejrua** [reply]
A query with "group by" "having" ?

**Reply by ganigeorgiev** [reply]
@ejrua Aggregation type of queries are supported only if you use PocketBase as framework and interact directly with the `app.Dao().DB()` instance.

**Reply by ejrua** [reply]
thanks for your answer, but I think adding "Custom query" to the Javascript SDK is necessary for this great tool

**Reply by ganigeorgiev** [reply]
Could you elaborate on your use case? `GROUP BY` and `HAVING` usually makes sense with aggregations, eg. like count, sum, etc.

You can't execute raw sql queries from client-side because there is no easy way to restrict what actually is being fetched from the database.

**Reply by ejrua** [reply]
thank, well, learn "GO"

**Reply by spydev** [reply]
`GROUP BY` and `HAVING` is not currently possible by using the SDK but can we do it with API itself? Am asking bzoc I have no experience in GO lang and trying PocketBase through JavaScript or using it in PHP by accessing API directly

**Reply by ganigeorgiev** [reply]
@spydev

> GROUP BY and HAVING is not currently possible by using the SDK but can we do it with API itself?

No, this is not possible at the moment without extending PocketBase. Eventually in the near future we may add support for "view" collection type and/or "raw sql" field type per https://github.com/pocketbase/pocketbase/issues/311#issuecomment-1312790397, but currently aggregations are supported only through custom db queries when using [PocketBase as framework](https://pocketbase.io/docs/use-as-framework/).
